source/sdk/lang/Exception.ooc

```diff
@@ -1,397 +1,395 @@
+	}
 }
 
-_takeStackFrame: func -> StackFrame {
-  exceptionStack get() as Stack<StackFrame> pop() as StackFrame
+ExceptionContext: class {
+	_exceptionJumpLabel: Int = 1
+	_localFrames := ThreadLocal<Stack<StackFrame>> new()
+	_currentException := ThreadLocal<Exception> new()
+	_stackFramesToDelete := Stack<StackFrame> new()
+	init: func
+	pushStackFrame: func -> StackFrame {
+		stack: Stack<StackFrame>
+		if (!this _localFrames hasValue?())
+			stack = Stack<StackFrame> new()
+		else
+			stack = this _localFrames get()
+		buf := StackFrame new()
+		stack push(buf)
+		this _localFrames set(stack)
+		buf
+	}
+
+	cleanupStackFrames: func {
+		while (!this _stackFramesToDelete isEmpty) {
+			frame := this _stackFramesToDelete pop() as StackFrame
+			if (frame)
+				frame free()
+		}
+	}
+
+	setException: func (exception: Exception) {
+		this _currentException set(exception)
+	}
+
+	getException: func -> Exception {
+		this _currentException get()
+	}
+
+	popStackFrame: func {
+		frame := this _localFrames get() as Stack<StackFrame> pop() as StackFrame
+		if (frame)
+			frame free()
+		this cleanupStackFrames()
+	}
+
+	takeStackFrame: func -> StackFrame {
+		frame := this _localFrames get() as Stack<StackFrame> pop() as StackFrame
+		this _stackFramesToDelete push(frame)
+		frame
+	}
+
+	hasStackFrame: func -> Bool {
+		this _localFrames hasValue?() && this _localFrames get() as Stack<StackFrame> size > 0
+	}
 }
 
-_hasStackFrame: inline func -> Bool {
-    exceptionStack hasValue?() && exceptionStack get() as Stack<StackFrame> size > 0
-}
+exceptionContext := ExceptionContext new()
 
-assert: extern func(Bool)
+_pushStackFrame: func -> StackFrame { exceptionContext pushStackFrame() }
+_popStackFrame: func -> StackFrame { exceptionContext popStackFrame() }
+_getException: func -> Exception { exceptionContext getException() }
 
 version(windows) {
-    import native/win32/[types, errors]
+	import native/win32/[types, errors]
 
-    getOSErrorCode: func -> Int {
-        GetLastError()
-    }
+	getOSErrorCode: func -> Int {
+		GetLastError()
+	}
 
-    getOSError: func -> String {
-        GetWindowsErrorMessage(GetLastError())
-        
-    }
+	getOSError: func -> String {
+		GetWindowsErrorMessage(GetLastError())
+	}
 } else {
-    errno: extern Int
-    strerror: extern func (Int) -> CString
+	errno: extern Int
+	strerror: extern func (Int) -> CString
 
-    getOSErrorCode: func -> Int {
-        errno
-    }
+	getOSErrorCode: func -> Int {
+		errno
+	}
 
-    getOSError: func -> String {
-        x := strerror(errno)
-        return (x != null) ? x toString() : ""
-    }
+	getOSError: func -> String {
+		x := strerror(errno)
+		return (x != null) ? x toString() : ""
+	}
 }
 
-raise: func(msg: String) {
-    Exception new(msg) throw()
+raise: func (msg: String) {
+	Exception new(msg) throw()
 }
 
-raise: func ~withClass(clazz: Class, msg: String) {
-    Exception new(clazz, msg) throw()
+raise: func ~withClass (clazz: Class, msg: String) {
```

Comment:
There are other examples. `class` is a keyword, so there may be problems with using it as a variable name (however, it can be used as an instance variable, so it may just be an old relic from a less civilized age...)